### DISCUSSION TITLE
Architectural advice please - multiple tables for a single model

### DISCUSSION THREAD

**Discussion by wickham-c** [filamentphp/filament#12288] [Help] [answered]
### Package

Panel builder

### Package Version

v3.2.52

### How can we help you?

I have a project which consists of a `Project` model and a `Task` model. In the `Project` model, I've defined a `HasMany` relationship named `tasks()` and a reciprocal `BelongsTo` relationship in the `Task` model named `project()`.

I've subsequently created a Filament resource named `ProjectResource` and a relation manager named `TasksRelationManager` and linked the `ProjectResource` to the relation manager via its `getRelations()` method. This is working superbly as I can see all of the tasks associated with a particular project. I'm very impressed with how straightforward this was to setup.

I'm at the point now where I need some advice on how to make further use of the `Task` model in other contexts within my Filament panel. I would like to...

1. Add a "**Tasks**" navigation item to the sidebar (just under the existing "Projects" item). When clicked, a page will be presented listings Tasks which have been assigned to the logged in user. This will save the user from having to dive into each individual project. Am I right in assuming that I need to create a `TaskResource` to facilitate this (rather than somehow reuse the existing `TasksRelationManager`)?
2. Add a "**Reports**" navigation item with a "**To Invoice**" sub-item to the sidebar (separate from the "Projects" and "Tasks" items). When "To Invoice" is clicked, a page will be presented listing invoiceable Tasks (i.e. any tasks within the entire company which have been marked with a _"to be invoiced"_ status). The Finance team will view this report to see which clients they need to send invoices to. They should be able to click on a table row to view a task in detail. This table will include quite a different set of columns compared to those displayed in the Tasks list described in item 1 above. What would be the best way to achieve this? Another new "resource" (e.g. `InvoiceableResource`), or reuse of the `TaskResource` created for item 1, or a new custom page with Livewire table added to it etc.?

What I essentially want is several different lists/tables to display Tasks in different contexts, the create / view / edit task forms can be common to all contexts. I'm loathe to heavily manipulate a single table (by showing/hiding columns) to meet the needs of each context, hence would like to know the _best way to have multiple tables presenting `Task` model instances in different ways_.

Any pointers would be gratefully received.

**Comment by micraux** [comment, marked as answer]
Hello,

Regarding your point 1, yes, you have to create a `TaskResource`, that will give you all the power of Filament to manage them.

For example, you could create new `Tasks` inside the `TasksRelationManager` reusing the form defined in your `TaskResource`.

Regarding your point 2, you may consider using the `Team feature`...

Do `Users` belong to many `Projects` and `Projects` have many `Users`?

**Reply by wickham-c** [reply]
Thanks for your advice @micraux 

Reusing forms defined in the `TaskResource` within the `TasksRelationManager` sounds like a good way forward.

A `Project` has a project manager (`User`) and each project `Task` has an assigned user... 

- In the `Project` model, I have defined a `BelongsTo` relationship with the `User` model named `projectManager`.
- In the `Task` model, a `BelongsTo` relationship is defined with the `User` model named `assignedTo`.

Is the "Team feature" you mentioned a Filament feature?

Currently, each `User` belongs to a `Department`.

**Reply by micraux** [reply]
Yes, I mentioned the [Multi-tenancy feature](https://filamentphp.com/docs/3.x/panels/tenancy) included in Filament.

Depending on the features covered by your application, you may want to use this feature to separate each `Department` _(do a Department manager have to manage his Users?_) or each `Project` (_can any User navigate through all Projects?_)...

For your Reports feature, you can create an `InvoiceableResource` based on the `Task` model, you can also create a new `Panel` dedicated to the Finance team (_do any User need to see the Report entry in the menu_?)

You might need a permission plugin:
- [Spatie Roles Permissions](https://filamentphp.com/plugins/tharinda-rodrigo-spatie-roles-permissions) (better for Multi-tenancy)
- [Shield](https://filamentphp.com/plugins/bezhansalleh-shield) (doesn't integrate Multi-tenancy out of the box)

**Reply by wickham-c** [reply]
Many thanks for the pointers and recommendations @micraux - I'll have to get familiar with the multi-tenancy feature and work out the best way forward for this particular app.

I also stumbled across this [section](https://filamentphp.com/docs/3.x/panels/resources/relation-managers#sharing-a-resources-form-and-table-with-a-relation-manager) in the Filament docs, which will likely prove useful in terms of form/table reuse.
